pwmig/libseispp/metadata.3
'\" te
.TH METADATA 3 "%G"
.SH NAME
Metadata - A C++ handler for variable metadata in an object
.SH SYNOPSIS
.nf
class Metadata
{
public:
        Metadata();
        Metadata(Metadata& md);
        Metadata operator=(const Metadata& md);
        ~Metadata();
        double get_metadata_by_name(string name);
        int get_metadata_by_name(string name);
        string get_metadata_by_name(string name);
        bool get_metadata_by_name(string name);
        void put_metadata_by_name(string name,double value);
        void put_metadata_by_name(string name,int value);
        void put_metadata_by_name(string name,string value);
        void put_metadata_by_name(string name,char *value);
        void put_metadata_by_name(string name,bool value);
        void load_metadata(string name);
};
void get_metadata_error_handler(int ierr) 
.fi
.SH DESCRIPTION
.LP
Metadata is a proverbial problem in data processing.  
By metadata I mean a set of parameters that are associated with 
a data object.  Some parameters are essential for describing
an object while others are not always needed.  The later 
are what I mean by metadata.  That is, one
algorithm may require a set of metadata while another does not.
Dealing with this causes major portability and consistency issues
in data processing.  The traditional solution is using a fixed 
name space through parameters stored as headers.  The more modern
approach would be to input required parameters and use some 
elegantly designed inheritance mechanism with an object-oriented
programming language like C++.  That is, each algorithm would 
inherit a base class and add other parameters to a data structure
it wanted to use internally.  
.LP
The Metadata class in this library is a generalization of this idea.
Rather than burden code with different sets of extracted metadata for
each algorithm, the model here is to place the metadata in a private
namespace.  This is a classic use of data hiding and abstraction of
the interface to a concept:  here the pool of metadata tagged with 
a name key.  The basic idea is that a data object to be used in 
some data processing model will inherit Metadata as a base class.
Truly required data from the Metadata space can (and should) be 
extracted immediately into binary components of the inherited 
class.  Other elements can be left in the Metadata space and 
extracted only when needed by a particular algorithm within a 
program.  This allows data objects to be specified with the 
minimum number of base parameters while having some assurance
that algorithms using other parameters have a hope of getting
access to them without creating a specialized data structure 
for each algorithm.  
.LP
The Metadata space is created by one of the constructors 
described in detail below.  The \fIload_metadata\fR function 
can be used to compile a block of parametric data into a 
Metadata space.  The current implementation uses the pf(5)
format of Antelope and calls pfcompile( see pf(3)) but 
the idea is that alternative formats could be used to build
a Metadata space provided the have a mechanism to tag each
item with a unique name.  Since parameters of standard types
can be placed into the Metadata space one at a time with the 
overloaded \fIput_metadata_by_name\fR functions.  The name
field in all puts is used to tag the parameter.  If the same
name is used multiple times the last put will be the one used.
(The follows the same model described in pf(3) on which the
library is based.  If anyone ever tried to replace that 
functionality with another approach this functionality should
be maintained, however, as applications will depend on it.)
.LP
Metadata are retrieved by the overloaded \fIget_metadata_by_name\fR
functions.  The get routines will throw an exception if the requested
parameter is not found in the Metadata space.  The function
\fIget_metadata_error_handler\fR is supplied with the interface for
convenience, but the user may and probably should write their own
error handler.  The reason is that only the requesting algorithm 
generally knows when a piece of metadata is essential or could
be set to some stock value and what that stock value should be.  
For this reason the \fIget_metadata_error_handler\fR does little
more than post an error message to cerr and return.  The only time
it will cause termination is if it returns an illegal error code
since that normally would mean something is seriously wrong with
the state of the running program.  In short, all calls to the
get functions should be surrounded by a try block and appropriate
catch handlers should deal with each error individually.  
.LP
The default parameter problem is so pervasive, on the other hand,
that the Metadata interface uses a more elaborate model for trying
to handle this problem in a general way.  On program startup a 
program using this library will search for an parameter file
with the default name of "metadata_defaults.pf".  (This can be
changed to a different parameter file by setting the environment
variable METADATA_DEFAULTS to an alternative pf name -- noting 
the usual rule about NOT including the ".pf" extension.)  
These parameters are loaded as defaults for the Metadata space.  
The default constructor loads this Metadata to build the defaults
for parameters.  This provides a general way to set defaults 
at run time, although it may cause serious confusion if the
user does not recognize that this is going on.  
.SH FILES
.LP
A defaults parameter file must be found in the PFPATH 
or a program using this library will die immediately.  
.SH ENVIRONMENT
.LP
METADATA_DEFAULTS can be used to redefine the name for the defaults
parameter file.  If it is null the default of "metadata_defaults" is
used.
.SH LIBRARY
$(STOCKLIBS)
.SH "SEE ALSO"
.nf
pf(3), pf(5), pfecho(1)
.fi
.SH "BUGS AND CAVEATS"
.LP
The dark side of the generality is some loss of efficiency and 
bloated memory usage because in this implementation all metadata
are stored as strings in an Antelope pf structure.  
.SH AUTHOR
.nf
Gary L. Pavlis
Indiana University
pavlis@indiana.edu
.\" $Id: metadata.3,v 1.1 2003/01/22 15:28:34 pavlis Exp $


prototypes/lib/libseispp/metadata.C
#include <iostream.h>
#include <string>
using namespace std;
#include "stock.h"
#include "pf.h"
#include "metadata.h"
//
// needed until antelope 4.5
//
Pf *pfdup(Pf*);
//
// This file is used to create a shared object library.  This 
// approach may not be universally applicable, but it is known to work
// in solaris.  When a library is first openned the _init() function is
// executed.  Here we utilized this to initialized a global parameter 
// space for the library used as a default template for metadata
//
static Pf *Metadata_defaults_pf;
void _init()
{
	char *mddname=(char *)"metadata_defaults";
	char *name_from_env;
	char *name;
	int ierr;
	// Get name from the environment
	// revert to default pf name if not defined 
	name_from_env = getenv("METADATA_DEFAULTS");
	if(name_from_env==NULL)
		name=mddname;
	else
		name=name_from_env;

	ierr = pfread(name,&Metadata_defaults_pf);
	if(ierr)
	{
		cerr << "Metadata_default initialization failed\npfread failed for" << name <<".pf\n";
		exit(-1);
	}
}
//
// The default constructor uses the defaults as a template which
// allows new parameters to replace defaults in place.
//
Metadata::Metadata()
{
	if(Metadata_defaults_pf==NULL)_init();
	pf=pfdup(Metadata_defaults_pf);
};
Metadata::~Metadata()
{
	pffree(pf);
};
Metadata::Metadata(const Metadata& md)
	throw(Metadata_parse_error)
{
	char *pfstr;  
	int ierr;

	// This approach uses the default as a template then compiles
	// the contents of md.pf on top of the defaults.  This is the
	// way in the current pf library to do this.  An ugly solution
	// in terms of memory because of the conversion to and from a string
	if(Metadata_defaults_pf==NULL)_init();
        pf=pfdup(Metadata_defaults_pf);
	pfstr = pf2string(md.pf);
	ierr = pfcompile(pfstr,&pf);
	// Just post this error and continue.  It should never happen, but
	// at least we handle the return from pfcompile correctly.
	if(ierr) throw Metadata_parse_error(ierr,
		"pfcompile error in copy constructor");
	free(pfstr);
}
Metadata& Metadata::operator=(const Metadata& md)
{
	if(&md!=this)
	{
		if(pf!=NULL) pffree(pf);
		pf=pfdup(md.pf);
	}
	return(*this);
}
//
// These functions get and convert values
//
double Metadata::get_double(string s)
	throw(Metadata_get_error)
{
	void *result;
	double val;
	int pftype_return;
	char *char_val;

	pftype_return = pfget(pf,(char *)s.c_str(),&result);
	if(pftype_return != PFSTRING) 
		throw Metadata_get_error("double",s,"");
	char_val=pfget_string(pf,(char *)s.c_str());
	val = atof(char_val);
	return(val);
}
int Metadata::get_int(string s)
	throw(Metadata_get_error)
{
	void *result;
	int val;
	int pftype_return;
	char *char_val;
	pftype_return = pfget(pf,(char *)s.c_str(),&result);
	if(pftype_return != PFSTRING) 
		throw  Metadata_get_error("int",s,"");
	char_val=pfget_string(pf,(char *)s.c_str());
	val = atoi(char_val);
	return(val);
}
string Metadata::get_string(string s) 
	throw(Metadata_get_error)
{
	void *result;
	string val;
	int pftype_return;
	char *char_val;
	pftype_return = pfget(pf,(char *)s.c_str(),&result);
	if(pftype_return != PFSTRING) 
		throw  Metadata_get_error("string",s,"");
	char_val=pfget_string(pf,(char *)s.c_str());
	val = char_val; //= is overloaded for this case so is a simple assign
	return(val);
}
bool Metadata::get_bool(string s)
	throw(Metadata_get_error)
{
	void *result;
	int val;
	int pftype_return;
	//
	// Boolean will default to false if not found in the pf
	// this is consistent with documentation and is rational behaviour
	//
	val = pfget_boolean(pf,(char *)s.c_str());
	if(val) 
		return(true);
	else 
		return(false);
}
Tbl *Metadata::get_list(string s)
	throw(Metadata_get_error)
{
	void *result;
	int val;
	int pftype_return;
	Tbl *t;
	pftype_return = pfget(pf,(char *)s.c_str(),&result);
	if(pftype_return != PFTBL) 
		throw  Metadata_get_error("Antelope Tbl",s,"");
	t = pfget_tbl(pf,(char *)s.c_str());
	return(t);
}
Arr *Metadata::get_map(string s)
	throw(Metadata_get_error)
{
	void *result;
	int val;
	int pftype_return;
	Arr *a;
	pftype_return = pfget(pf,(char *)s.c_str(),&result);
	if(pftype_return != PFARR) 
		throw  Metadata_get_error("Antelope Arr",s,"");
	a = pfget_arr(pf,(char *)s.c_str());
	return(a);
}

//
// Functions to put things into metadata object
//
void Metadata::put_metadata(string name, double val)
{
	pfput_double(pf,(char *)name.c_str(),val);
}
void Metadata::put_metadata(string name, int val)
{
	pfput_int(pf,(char *)name.c_str(),val);
}
void Metadata::put_metadata(string name, string val)
{
	pfput_string(pf,(char *)name.c_str(),(char *)val.c_str());
}
// for C style strings, we should not depend on the compiler
void Metadata::put_metadata(string name, char *val)
{
	pfput_string(pf,(char *)name.c_str(),val);
}
void Metadata::put_metadata(string name, Arr *val)
{
	pfput_arr(pf,(char *)name.c_str(),val);
}
void Metadata::put_metadata(string name, Tbl *val)
{
	pfput_tbl(pf,(char *)name.c_str(),val);
}
void Metadata::put_metadata(string name, bool val)
{
	if(val)
		pfput_boolean(pf,(char *)name.c_str(),1);
	else
		pfput_boolean(pf,(char *)name.c_str(),0);
}
void Metadata::load_metadata(string mdin) 
	throw(Metadata_parse_error)
{
	int ierr;
	// We might think this was needed:  if(pf!=NULL) pffree(pf);
	// it is not because pfcompile checks for *pf==NULL and
	// assumes it is valid and to be updated if nonzero.
	// This is a handy way to deal with defaults
	// The char * cast is necessary to keep the compiler
	// from bitching about a const char
	ierr = pfcompile((char *)mdin.c_str(),&pf);
	if(ierr!=0) throw Metadata_parse_error(ierr,"Failure in load_metadata");
}
// near dup of above done for convenience
void Metadata::load_metadata(char *mdin) 
	throw(Metadata_parse_error)
{
	int ierr;
	// We might think this was needed:  if(pf!=NULL) pffree(pf);
	// it is not because pfcompile checks for *pf==NULL and
	// assumes it is valid and to be updated if nonzero.
	// This is a handy way to deal with defaults
	ierr = pfcompile(mdin,&pf);
	if(ierr!=0) throw Metadata_parse_error(ierr,"Failure in load_metadata");
}

//
//Sometimes we need to not copy all of the metadata from one object
//to another.  This function allows selective copy driven by a list
//

void  copy_selected_metadata(Metadata& mdin, Metadata& mdout,
		                list<Metadata_typedef>& mdlist)
	throw(Metadata_error)
{
	list<Metadata_typedef>::iterator i;
	Metadata_typedef& mdti=*i;
	int count;

	for(i=mdlist.begin(),count=0;i!=mdlist.end();++i,++count)
	{
		MDtype mdtest;
		double r;
		int iv;
		string s;
		Tbl *t;
		Arr *a;
		bool b;

		mdtest = mdti.mdt;
		try {
			switch(mdtest)
			{
			case MDreal:
				r=mdin.get_double(mdti.tag);
				mdout.put_metadata(mdti.tag,r);
				break;
			case MDint:
				iv=mdin.get_int(mdti.tag);
				mdout.put_metadata(mdti.tag,iv);
				break;
			case MDstring:
				s=mdin.get_string(mdti.tag);
				mdout.put_metadata(mdti.tag,s);
				break;
			case MDlist:
				t=mdin.get_list(mdti.tag);
				mdout.put_metadata(mdti.tag,t);
				break;
			case MDmap:
				a=mdin.get_map(mdti.tag);
				mdout.put_metadata(mdti.tag,a);
				break;
			case MDboolean:
				b=mdin.get_bool(mdti.tag);
				mdout.put_metadata(mdti.tag,b);
				break;
			default:
				cerr<<"Fatal: copy_selected_metadata was passed illegal type definition\nFatal error as this indicates a coding error that must be fixed" << endl;
				exit(-1);
			};
		} catch( Metadata_error merr)
		{
			cerr << "Error in copy_selected_metadata at item ";
			cerr << count << "with tag" << mdti.tag <<"\n" ;
			cerr << "Copy truncated" << endl;
			merr.log_error();
			throw;
		}
	}
}
// Simple function to dump all metadata in a md object to 
// stdout
//
void Metadata::print_all_metadata()
{
	char *md_contents;
	md_contents=pf2string(pf);
	if(md_contents==NULL)
		cout << "ERROR: Metadata object is empty";
	else
	{
		cout << md_contents;
		free(md_contents);
	}
}
